### DISCUSSION TITLE
DataGrid DataGridTemplateColumn Button SelectedItem null if row is not selected

### DISCUSSION THREAD

**Discussion by evilbaschdi** [AvaloniaUI/Avalonia#17167] [Q&A]
### Describe the bug

I'm not sure, if is really a bug in the code or if I do something wrong, but I need help here :-)

I've migrated some WPF Application to Avalonia some time ago. Both using MVVM
In both worlds I've hosted a Button inside a DataGrid Column to start an action here.

Now the issue is that the SelectedItem of the DataGrid is NULL if I just hit that Button in case I do not select the containing row before.
But! working with the NumericUpDown inside the same row makes not problem.
So maybe there's some problem while defining my Command

### To Reproduce

**Repository:** [https://github.com/evilbaschdi/VmMachineHwVersionUpdater](https://github.com/evilbaschdi/VmMachineHwVersionUpdater) (develop branch contains a debug zip file)
**Project:** VmMachineHwVersionUpdater.Avalonia

**Extracting vmware.zip for Grid Data:**

1. The "res" directory contains a "vmware.zip" file. Extract its contents to "C:\vm\VMware" (defined in the Core project's VmPools.json file).
2. Look for the "Button Column Definition" starting at Line 40 (Header="Start") in the Views\MainWindow.axaml file.
3. The MainWindowViewModel is referenced in App.axaml.cs and has a public property "SelectedMachine" that is assigned to the DataGrid's SelectedItem.

**Application Functionality:**

- The application searches for ".vmx" files within defined "vmpool" locations.
- It reads the contents of these files and their directories.
- The results are displayed in the DataGrid.

**Using the DataGrid:**

- The DataGrid allows you to perform actions on individual entries.
- One action is "powering on" a virtual machine using the "Start" button.

**SelectedMachine and ICurrentItem:**

- Other columns and context menu items in the grid rely on the "SelectedMachine" property. 
  They assign the current item to "ICurrentItem" without requiring explicit row selection.
- However, clicking the "Start" button **does not** automatically set "SelectedMachine". This causes "ICurrentItem.Value" to be null, preventing the application from functioning correctly.

### Expected behavior

- Click the Button
- Get current row
- Assign the selected Item correctly

As already mentioned before; I'm not 100% sure if this an issue of Avalonia. Maybe it also could be that my Buttom.Command is defined wrong.

### Avalonia version

11.1.1

### OS

Windows

**Comment by timunie** [comment]
Please try to make it a minimum sample to reproduce. If you need help for a complete App, please request paid support. Thanks for your understanding.

**Comment by evilbaschdi** [comment]
The problem also occurs with "public Country SelectedCountry" in

https://github.com/evilbaschdi/EvilBaschdi.Core.Avalonia/blob/develop/src/EvilBaschdi.Core.Avalonia.DummyAppMvvm/ViewModels/ExtendedInformationViewModel.cs
I also already implemented a Button on the very right of each row that will open an new window including the information if the current value is NULL or not

is that minimum enough?

**Comment by timunie** [comment]
From a first view it looks like you never raise PropertyChanged when you update SelectedContry. Thus the UI will not know about the new selection made. 

> [!TIP]
> Do you have a basic understanding about MVVM? If not I suggest you check out Avalonia.Samples about mvvm usage.

**Comment by evilbaschdi** [comment]
Even if I do raise RaisePropertyChanged inside the setter of the SelectedItem its not working because that setter never gets called.
It's only called while selecting the grid row before.

**Comment by timunie** [comment]
it is depended on several other projects and even a feed that I don't know. I cannot check all of the code if it is save to run, which I have to do to debug it on my own. What I can think of is that Button will raise the command before DataGrid gets the possibility to select a row. I suggest to add a CommandParamter to your Command which you simply bind to the rows DataContext. 

Here is a working sample, maybe you can check out what I did and adopt it.

[DataGridWithButton.zip](https://github.com/user-attachments/files/17186202/DataGridWithButton.zip)

**Comment by evilbaschdi** [comment]
thx for your example but its implemented with that CommunityToolkit. But I (still?) Use ReactiveUI

**Reply by timunie** [reply]
Feel free to use RxUI. The mvvm Framework is irrelevant to this issue.
